### PR TITLE
Fix `num_items_in_batch` over-counting for causal LM losses

### PR DESCRIPTION
**Pull request by qgallouedec**
## The bug

`Trainer._get_num_items_in_batch` counts labels at every position. But for causal LM the loss shifts labels — position 0 is never a prediction target. The denominator is too large by **`num_rows` per micro-batch**, systematically under-scaling every causal LM loss and gradient.

```
labels (trainer counts these):           shift_labels = labels[..., 1:]
                                         (what ForCausalLMLoss uses)
┌────┬────┬────┬────┬────┐                    ┌────┬────┬────┬────┐
│ t0 │ t1 │ t2 │ t3 │ t4 │  count = 5         │ t1 │ t2 │ t3 │ t4 │  count = 4
└────┴────┴────┴────┴────┘                    └────┴────┴────┴────┘
  ↑
  position 0 is dropped by the shift — over-count = 1 per row
```

```python
shift_labels = labels[..., 1:]           # ← numerator: 4 CE terms
loss = sum_ce / num_items_in_batch       # ← denominator: counted 5
```

`loss_type` is set on every `PreTrainedModel` (see `modeling_utils.py`). Only causal LM loss types are touched; `ForMaskedLM`, classification, etc. are unaffected.

Reported causal LM loss becomes slightly larger (correctly so, the denominator was too big). Shift is `num_rows / total_tokens` per step. Gradient magnitudes scale by the same factor.

## Tests

All `TrainerGradientAccumulationTest` tests pass (4/4). A padded vs padding-free reproducer now matches with Δgrad_norm = 0.

## How it surfaced

TRL's invariance suite compared SFT with `padding_free=False` vs `padding_free=True`. Loss curves almost matched, but `grad_norm` drifted with a systematic +0.17 bias over 50 steps. The padding-free collator masks `labels[position_ids == 0] = -100`, which *incidentally* matches the post-shift count, so it exposed the padded path's over-count.

### CODE REVIEW

File: src/transformers/trainer.py [hunk 1]
```diff
@@ -508,6 +508,14 @@ def __init__(
         self.label_names = default_label_names if self.args.label_names is None else self.args.label_names
         self.can_return_loss = can_return_loss(model_to_inspect.__class__)
 
+        # Causal LM losses shift labels internally (predictions at position i target label[i+1]), so position 0 of
+        # each row is never a prediction target. The valid-prediction count used by `num_items_in_batch` must therefore
+        # be taken over `labels[..., 1:]`, not the full label tensor
+        self._loss_shifts_labels = getattr(model_to_inspect, "loss_type", None) in (
```
Comment (vasqu) [inline src/transformers/trainer.py]: I think this is too simple. I think if we were to go that way we should inspect the actual loss function, e.g. `CsmForConditionalGeneration` would fail here, no?

But, my bigger issue is that this covers the most simple usecase where we only prepare the tokenized input and pass that but what if we were to use a data collator that properly prepares the shifted labels? We would now count 1 too much

But definitely needed to fix in general!

**Comment by vasqu** [comment]
Also definitely need a test that covers this edge case

File: src/transformers/trainer.py [hunk 2]
```diff
@@ -508,6 +508,14 @@ def __init__(
...
+            "ForCausalLM",
+            "ForConditionalGeneration",
+        )
+
```
Comment (SunMarc) [inline src/transformers/trainer.py:518]: as @vasqu pointed, maybe we can have the following so that it is a bit more robust. 
```python
from transformers.loss.loss_utils import LOSS_MAPPING, ForCausalLMLoss
    self._loss_shifts_labels = (
        LOSS_MAPPING.get(getattr(model_to_inspect, "loss_type", None)) is ForCausalLMLoss
    )
```

File: src/transformers/trainer.py [hunk 3]
```diff
@@ -2133,7 +2141,13 @@ def _get_num_items_in_batch(self, batch_samples: list, device: torch.device) ->
         if count_num_items_in_batch:
             # For now we don't support object detection
             try:
-                num_items_in_batch = sum((batch["labels"].ne(-100)).sum() for batch in batch_samples)
+                # Causal LM losses shift labels; count over `labels[..., 1:]` to avoid over-counting position 0.
+                labels_for_count = (
+                    [batch["labels"][..., 1:] for batch in batch_samples]
+                    if self._loss_shifts_labels
+                    else [batch["labels"] for batch in batch_samples]
+                )
```
Comment (SunMarc) [inline src/transformers/trainer.py]: Maybe to fix @vasqu point, we can also take into account the case where `shift_labels` is prepared by the user ? 

```python
  labels_for_count = [
      batch["shift_labels"] if "shift_labels" in batch
      else batch["labels"][..., 1:] if self._loss_shifts_labels
      else batch["labels"]
      for batch in batch_samples
  ]
```

**Review by SunMarc — COMMENTED** [review]
Thanks, left some minor comments but happy to merge it in general. Can you add a small test to check that we are calculating correctly the `num_item_per_batch` in case we have `ForCausalLM` ?

**Review by SunMarc — APPROVED** [review]
Thanks !

**Comment by vasqu** [comment]
Seems like we have 1 new failure https://app.circleci.com/pipelines/github/huggingface/transformers/175911/workflows/b1032ab4-41a0-44f1-b1cd-072bad362f4d/jobs/2326963

Thanks tho, overall LGTM as well

**Comment by qgallouedec** [comment]
> Seems like we have 1 new failure

https://github.com/huggingface/transformers/pull/46204/commits/0a3d375c57e42ace4986c73953c6ff77968b7ead should fix it
